tools/codegen/gen_functionalization_type.py

```diff
@@ -195,7 +195,22 @@ def emit_inplace_functionalization_body(
                     if arg.type.is_tensor_like() else None,
         f.func.arguments.flat_all))
 
-    if functional_op is None:
+    # Note [functionalizating copy_() and not preserving strides]
+    # copy_() can't be functionalized, since there doesn't exist an out-of-place variant.
+    # We could add one, but that would be sub-optimal for functorch: copy() would need to allocate a fresh tensor.
+    # This may seem like a large hack for one optimization, but copy_() is one of the most common inplace operators.
+    # Instead, we can replace `self.copy_(src)` with `src.to(self).expand_as(self)`.
+    # This maintains the exact same semantics, EXCEPT that we don't preserve the strides from `self`.
+    # This seems like a reasonable tradeoff, for a few reasons:
+    # - mutation removal is only used by functorch, and not by Vulkan or XLA. Functorch already doesn't preserve strides.
+    # - There are actually a few other places where the functionalization pass currently doesn't support strides:
+    #   calls to slice/diagonal_scatter don't currently preserve the strides of their inputs (but maybe we should fix this).
+    if str(f.func.name) == 'copy_':
+        exprs = [keyset] + [a.name for a in unwrapped_args_ctx]
+        functional_call_str = f"""\
+            auto tmp_intermediate = at::_ops::to_other::redispatch({keyset}, src_, self_, non_blocking, false, c10::nullopt);
+            tmp_output = at::_ops::expand_as::redispatch({keyset}, tmp_intermediate, self_);"""
```

Comment:
There are some interesting cases that we should test.
- A copy_ case where the dtype and shape are the same (this is already present in the functionalization PR)
- A copy_ case where the dtype is different
- A copy_ case where the shape is different
- Because `expand_as` returns a view and an expanded tensor, we should check that further mutation (in the original code) of the result still produces the correct output.

So, for example:
```
def f(x):
  tmp = torch.zeros(2, 2)
  tmp_slice = tmp.diagonal()
  r1 = tmp_slice.copy_(x)
  r2 = r1.add_(x)
  return x, r1, r2

cases = [
  torch.tensor(1.),
  torch.tensor(1, dtype=torch.long),
  torch.randn(2),
]
```

Reply:
oof thanks for pointing out https://github.com/pytorch/pytorch/pull/66101#discussion_r743713239, and that also pretty clearly points out the extra coverage for `copy_()` that you're suggesting (in some of your other examples, I **would** expect `to` to show up in the LoggingTensor trace). I'll add those in before landing the stack.